Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs

```diff
@@ -165,26 +203,43 @@ public override void OnActionReceived(float[] vectorAction)
         bpDict[forearmR].SetJointStrength(vectorAction[++i]);
     }
 
+    //Update OrientationCube and DirectionIndicator
+    void UpdateOrientationObjects()
+    {
+        worldDirToWalk = walkDirectionMethod == WalkDirectionMethod.UseTarget
+            ? target.position - hips.position
+            : worldDirToWalk;
+        m_OrientationCube.UpdateOrientation(hips.position, worldDirToWalk);
+        m_DirectionIndicator.MatchOrientation(m_OrientationCube.transform);
+    }
+
     void FixedUpdate()
     {
-        var cubeForward = orientationCube.transform.forward;
-        orientationCube.UpdateOrientation(hips, target.transform);
+        UpdateOrientationObjects();
+
+        var cubeForward = m_OrientationCube.transform.forward;
+
         // Set reward for this step according to mixture of the following elements.
-        // a. Velocity alignment with goal direction.
-        var moveTowardsTargetReward = Vector3.Dot(cubeForward,
-            Vector3.ClampMagnitude(m_JdController.bodyPartsDict[hips].rb.velocity, maximumWalkingSpeed));
-        if (float.IsNaN(moveTowardsTargetReward))
+        // a. Match target speed
+        //This reward will approach 1 if it matches perfectly and approach zero as it deviates
+        var matchSpeedReward = GetMatchingVelocityInverseLerp(cubeForward * walkingSpeed, GetAvgVelocity());
+        
+        //Check for NaNs
+        if (float.IsNaN(matchSpeedReward))
         {
             throw new ArgumentException(
                 "NaN in moveTowardsTargetReward.\n" +
                 $" cubeForward: {cubeForward}\n"+
                 $" hips.velocity: {m_JdController.bodyPartsDict[hips].rb.velocity}\n"+
-                $" maximumWalkingSpeed: {maximumWalkingSpeed}"
+                $" maximumWalkingSpeed: {m_maxWalkingSpeed}"
             );
         }
 
-        // b. Rotation alignment with goal direction.
-        var lookAtTargetReward = Vector3.Dot(cubeForward, head.forward);
+        // b. Rotation alignment with target direction.
+        //This reward will approach 1 if it faces the target direction perfectly and approach zero as it deviates
+        var lookAtTargetReward = (Vector3.Dot(cubeForward, head.forward) + 1) * .5F;
+
+        //Check for NaNs
         if (float.IsNaN(lookAtTargetReward))
```

Comment:
Can this really be NaN ? I don't think Vector3.Dot can produce a NaN.

Reply:
This was added to help debug the NaN issues we were having with Walker training. We can take these checks out once that is resolved